crypto/evp/p5_crpt.c

```diff
@@ -49,16 +56,14 @@ int PKCS5_PBE_keyivgen(EVP_CIPHER_CTX *cctx, const char *pass, int passlen,
     }
 
     ivl = EVP_CIPHER_iv_length(cipher);
-    if (ivl < 0 || ivl > 16) {
+    if (ivl != PKCS5_PBES1_KEY_IV_LENGTH) {
```

Comment:
You have changed the values here from 16 to 8 and also wired in that key and iv length must be the same and must be 8 bytes which isn't what the code had before. Why?

Reply:
The PBES1 algorithm defined in https://datatracker.ietf.org/doc/html/rfc2898#section-6.1 states:

>       3. Separate the derived key DK into an encryption key K consisting
>          of the first eight octets of DK and an initialization vector IV
>          consisting of the next eight octets:
> 
>                  K   = DK<0..7> ,
>                  IV  = DK<8..15> .

This is the constraint I used to determine the acceptable key/IV lengths.

Reply:
And you are applying this in a way that alters the logic - as these are properties of the input cipher.
You are now wiring them to a fixed value which is not what the code did previously. 

It matters little as to what the RFC states for things like this - we have a routine that can operate using any cipher and you are changing the logic of how it processes the input cipher significantly - and that should not be happening in this sort of code update.

Reply:
Tim are you aware of an actual use case?  Or is this a theoretical argument? I realize this is a public interface.. (albeit a fairly crappy one)

Reply:
It is a public function that takes and cipher and people do unusual things and wiring this to only work for iv==8 and keylength==8 seems rather inappropriate when the PR isn't about reducing the functionality to just what the RFC states.
We don't claim to limit it and 

If we wanted to do that then it should have been explicitly noted and raised and signed off by the OTC - not just mixed up in the middle of a patch that is meant to just be moving code around.

It really stood out as an input parameter validation change that had nothing to do with the patch.

Looking at the history some asserts got turned into these checks in 2019 a05bf83c7964bb3928b323fe356b9f70f105036d and the original asserts from 2002 are in 54a656ef081f72a740c550ebd8099b40b8b5cde0

There also appears to have been RC4 128-bit usage in the code originally (although not in the more recent code).

Reply:
Ok, sure, I can see that this being a public function restricts how much we can clean up the implementation.

I've re-introduced the original logic now.

Reply:
Thanks.

Reply:
If I remember rightly I think we did this initially because it was trashing over bits of the buffer that it shouldnt if the size was > 8.. Oh well..

Reply:
That recollection would be in error. The code is not "trashing over bits of the buffer". If it did then that should be separately fixed. If we wanted to change the behaviour that should be separately fixed. When moving code around it is not the time to be altering behaviours like this as that makes for a challenge in terms of tracking changes.

The code was written in a manner that allowed for non 8-byte cipher algorithm use - there is evidence of that usage in the past - and that isn't covered in the RFC - but that isn't all that unusual with this RFCs which were written after-the-fact and are often simplifications or codifications of implementations that were more rigid that necessarily the practice out there in the wild.

The code has an explicit range check for both keylength and ivlength both of which are safe for the entire range of the check if you read the code. 

Although I see a separate potential issue - if the IV length is 16 bytes then we will use the key as the IV as well which is probably not exactly the expected behaviour. But there is no "trashing" that I can see. If you think there is then there should be a separate issue called out for this and it should be resolved independent of this PR.

Reply:
Yep. I didnt quite mean a buffer oveflow :)
md_tmp + (16 - ivl) is just wrong with these constraints..
The keylen constraint is the size of the buffer (64 bytes) - so there would be no space for iv in this case.